### PR TITLE
Refs #21312 - limit locales bundled from intl packages

### PR DESCRIPTION
**Pull request by tstrachota**
Restricts locale files in packages `intl` and `react-intl` only to the subset of languages supported in the Foreman to avoid huge webpack bundles.

### CODE REVIEW

**Comment by tstrachota** [comment]
@ares @tbrisker may I ask you for the review?

**Comment by ohadlevy** [comment]
@tstrachota whats the size difference ? also note that currently we only load the active language vs all the languages at the same time, do you think we can  break it down to each language and load that conditionally?

File: config/webpack.config.js
Comment (ohadlevy) [inline config/webpack.config.js]: is there an easy way to maintain this list (instead of hard code?)
also, any idea why the intl js code is so big? which kind of strings are in it to begin with?

**Comment by tbrisker** [comment]
in development this reduced the bundle size from 68MB to 20MB, didn't test production. This is a big save any case, but I still think it may be good if possible to split these from the main bundle and build a bundle for every language so that the users will only need to download one language instead of 14, similar to what we do for legacy translations.

File: config/webpack.config.js
Comment (tstrachota) [reply config/webpack.config.js]: We could probably `ls ./locale/` and use names of the directories there.

Comment (tbrisker) [reply config/webpack.config.js]: perhaps get a list of subfolders in `locale/` and join by `|`?

**Comment by tstrachota** [comment]
I updated the webpack config to detect locales from `./locale/`.

I agree with adding per-language chunks. I'll need to invest more time into playing with how to do that though. So we might want to merge this improvement first and work on splitting the bundle with priority but independently.

File: config/webpack.config.js
Comment (tbrisker) [inline config/webpack.config.js]: This produces duplicates for languages that appear twice such as `en` or `zh`

Comment (tstrachota) [reply config/webpack.config.js]: Good catch, updated

**Comment by ares** [comment]
I guess we're close so we could postpone reverting deadline for tomorrow. Branching won't happen tonight based on my info.

**Comment by ares** [comment]
So we had a discussion with @tstrachota and here's our conclusions:

1) using intl for date/time format localization is long term a good idea, supported by all modern browsers, intl library that causes the big bundle is only used for legacy browsers and with this patch, we only bundle date strings which adds "only" 0.5 MB to ~1 MB we had before intl was merged, right now, the bundle.js in develop has 18.6 MB for vendor.js, see more info at https://gist.github.com/tstrachota/7beca280b3fa944463df1e434902bec2
2) the internationalization of strings such as "seconds", "months", "ago" is provided by modern browser intl API and it sounds reasonable to rely on that, not trying to mix it into our gettext
3) further internationalization of JS stack is subject of discussion, so far the plan is to continue extracting strings using `__()` but a discussion about other solutions could be held, it's orthogonal to this PR, my personal opinion is that no matter what we chose, the native intl browser support should be used where applicable
4) loading of just the language that is required would be good addition, if you think it's a hard dependency of this, we should revert the original PR as it will take longer time, if you're fine with adding this later, we'd prefer to merge this PR and keep dates components in for 1.17

@ohadlevy @tbrisker please let us know what you think and what you prefer in terms of point 4

**Review by ohadlevy — CHANGES_REQUESTED** [review]
@tstrachota can you please look into extracting the intl package to its own chunk? this will allow us conditionally to load the js file instead of always loading it into memory on all browsers.

**Comment by tstrachota** [comment]
I added a commit with prototype of including intl and its translations dynamically from separate chunks.

But I'm struggling with getting the tests run. I'm getting "SyntaxError: Unexpected token import" :-/
Webpack compilation works fine.

**Comment by ohadlevy** [comment]
import is an es6 syntax that webpack support,I'm guessing that you would
need to find the similar jest plugin to enable it for testing?

**Comment by tstrachota** [comment]
Ehm, ok, it was jest not updating cached `.babelrc`... should be ok now.

**Comment by ares** [comment]
What's the status now?  Is this ready or should we revert the original commit to unblock nightlies?

**Comment by ohadlevy** [comment]
@ares the tests currently fail, not sure how far we are at but it does seem pretty close

**Comment by tstrachota** [comment]
Although it looked promising and webpack built separate chunks for intl, I hit another issue. The intl chunk doesn't get loaded (hence the failing tests).
When I tested it I found out that import actually errors out with:
```
Uncaught (in promise) TypeError: Cannot read property 'call' of undefined
    at __webpack_require__ (bootstrap d2619090cf6a83b9d987:712)
    at fn (bootstrap d2619090cf6a83b9d987:117)
    at <anonymous>
```
I experimented with various dynamic loading styles (`require.ensure()`, `require()`) but with the same result :-/ 
So, no luck with the dynamic imports yet. We can still merge at least the first commit with the contexts that reduce the bundle size.

**Comment by mmoll** [comment]
On my x86_64 machine, the first commit reduced the memsize for the respective nodejs process to about 40%, that should suffice to let the DEB builds on 32 bit archs succeed again.

**Comment by ares** [comment]
@ohadlevy what is your preference? Thanks

**Comment by ohadlevy** [comment]
@ares @tstrachota I don't think it make a lot of sense to include in 1.17, I prefer we revert it and potentially reapply the smaller version to develop after branching 1.17-stable.
I would also think it make sense to consider revisiting moment.js for date / time as its a standalone library that is translated well.

**Comment by tstrachota** [comment]
I'm closing this PR. I'll include the code in the new one that I'm going to open.

**Pull request closed without merging**
